traversal/procedure/GraphProcedure.java

```diff
@@ -214,74 +150,160 @@ public boolean equals(Object o) {
         if (this == o) return true;
         if (o == null || getClass() != o.getClass()) return false;
         GraphProcedure that = (GraphProcedure) o;
-        return vertices.equals(that.vertices) && Arrays.equals(edges, that.edges) && startVertex().equals(that.startVertex());
+        return Arrays.equals(vertices, that.vertices);
     }
 
     @Override
     public int hashCode() {
-        int result = Objects.hash(vertices, startVertex);
-        result = 31 * result + Arrays.hashCode(edges);
-        return result;
+        return Arrays.hashCode(vertices);
     }
 
     @Override
     public String toString() {
         StringBuilder str = new StringBuilder();
         str.append("Graph Procedure: {");
-        List<ProcedureEdge<?, ?>> procedureEdges = Arrays.asList(edges);
-        procedureEdges.sort(Comparator.comparing(ProcedureEdge::order));
-        List<ProcedureVertex<?, ?>> procedureVertices = new ArrayList<>(vertices.values());
-        procedureVertices.sort(Comparator.comparing(v -> v.id().toString()));
-
-        str.append("\n\tvertices:");
-        for (ProcedureVertex<?, ?> v : procedureVertices) {
-            str.append("\n\t\t").append(v);
-        }
-        str.append("\n\tedges:");
-        for (ProcedureEdge<?, ?> e : procedureEdges) {
-            str.append("\n\t\t").append(e);
+        for (int i = 0; i < vertexCount(); i++) {
+            ProcedureVertex<?, ?> vertex = vertices[i];
+            str.append("\n\t").append(vertex);
+            for (ProcedureEdge<?, ?> edge : vertex.ins()) {
+                str.append("\n\t\t\t").append(edge);
+            }
```

Comment:
change the `toString()` to print out the vertex ordering, and group edges below each vertex that lead into that vertex as a group.